File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/ui/ScreensUseTemplates.kt
package com.perrystreet.woof.konsist.ui

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.KonsistUtils.body
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class ScreensUseTemplates : BehaviorSpec() {
    init {
        Given("A Screen composable") {
            val screens = KonsistUtils.screenComposables

            Then("It renders a Template or delegates to another Screen") {
                screens.assertTrue(message = Message) { screen -> TemplateOrScreenCallRegex.containsMatchIn(screen.body) }
            }
        }
    }

    private companion object {
        private val TemplateOrScreenCallRegex = Regex("""\b(Template\w+|[A-Z]\w*Screen)\(""")

        private val Message = LintRuleMessage(
            rule = "Screen composables use a Template* as their root layout.",
            why = "Templates own page structure (top bar, insets, paddings, scrolling). Screens only fill their slots, so every screen shares the same skeleton.",
            howToFix = "Wrap the screen content in an existing Template, or add a new Template to the design system.",
            badExample = "@Composable fun GridScreen(state: State) { Scaffold { LazyVerticalGrid { ... } } }",
            goodExample = "@Composable fun GridScreen(state: State) { TemplateGrid(topBar = { OrgNavigationHeaderBranded() }) { ... } }",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/UI/ScreensUseTemplates.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class ScreensUseTemplates: QuickSpec {
    override class func spec() {
        Given("A Screen view") {
            let screens = WoofHarmonize.screenViews

            Then("It renders a Template or delegates to another Screen") {
                screens.assertTrue(rule: rule) { $0.description.containsMatch(of: templateOrScreenCallPattern) }
            }
        }
    }

    private static let templateOrScreenCallPattern = try! NSRegularExpression(pattern: #"\b(Template\w+|[A-Z]\w*Screen)\("#)

    private static let rule = Rule(
        description: "Screen views use a Template* as their root layout.",
        rationale: "Templates own page structure (top bar, insets, paddings, scrolling). Screens only fill their slots, so every screen shares the same skeleton.",
        fixHint: "Wrap the screen content in an existing Template, or add a new Template to the design system.",
        badExample: "struct GridScreen: View { var body: some View { ScrollView { LazyVGrid { ... } } } }",
        goodExample: "struct GridScreen: View { var body: some View { TemplateGrid(topBar: { OrgNavigationHeaderBranded() }) { ... } } }"
    )
}
